### DISCUSSION TITLE
Blocked on getting receipts + listings

### DISCUSSION THREAD

**Discussion by bijoutrouvaille** [etsy/open-api#1000] [Q&A] [answered]
My client requires syncing shop receipts with our database. Some users have thousands of receipts they need to sync. 

First, I `getShopReceipts` 100 at a time. Second, I need to associate each one with its listings. If I just `getListingsByShopReceipt` I quickly reach the requests-per-second quota. If I `getListingsByShop` I have no way to associate the listings with the receipts as the Receipt object does not come with listings. I don't know how to proceed.

- I can't make the users click the button 300 times instead of 30
- client is running on AWS Lambas, which time out, and building a special queue system just for this is prohibitively expensive

How can work around this issue?
How come there's not an include parameter on `getShopReceipts`?

**Comment by Gareth064** [comment]
Hi @bijoutrouvaille, I would be happy to help figure this one out for you.

First I need to understand what data you want to get from Etsy into your system.
So far it looks like you want the Receipt and the listing it is associated with?
What about Listing Images?

**Reply by Gareth064** [reply, marked as answer]
OK, not knowing how many orders a single use might have, that you would want to sync on go. Lets just big a large number of 9000.

So you have 9000 Receipts to get, that is 90 requests to Etsy.

You said that you have no way of associating the listing to the Receipt, if you were to get the listings separately, but the Receipt has transactions on it, and the transactions have the ListingId to allow you to match the receipts items to the listings.

So you can either use `getListingsByShop` to get all the shops listings, including ones that aren't related to any receipts. Or you can get all your receipts first, then gather all the ListingId's from the transactions, get a distinct list of IDs, batch them up by 100, then use `getListingsByListingIds` passing 100 Listing Ids at a time.

Doing the above means you won't get anywhere near the 10,000 rate limit (per shop).

Note, if you want the images of the listings too then the getListings endpoint have includes on them for Images amongst other things.

If, we are talking 100,000's of receipts, per shop, then probably a better approach would be to build an csv import feature into your app. Each shop owner can then download all their receipts and listings using Etsy's easy to use export feature on their website, to get a csv so import to your app. This would work to allow a shop owner to bulk import their sales history from etsy. Then you use the Esty API to sync new and updated data from their onwards.

**Reply by bijoutrouvaille** [reply]
I haven't tried it yet, but that sounds like the solution I need.

What confused me is that you can get listings by receipt, when the underlying schema is actually `Receipt.Transactions.Listings`.

Thank you for your help @Gareth064

**Reply by Gareth064** [reply]
No worries. Pop back if you have any other questions.
